hare/algorithm.go

```diff
@@ -844,49 +851,57 @@ func (proc *consensusProcess) shouldParticipate(ctx context.Context) bool {
 	res, err := proc.oracle.IsIdentityActiveOnConsensusView(ctx, proc.signing.PublicKey().String(), types.LayerID(proc.instanceID))
 	if err != nil {
 		logger.With().Error("should not participate: error checking our identity for activeness",
-			log.Err(err), types.LayerID(proc.instanceID))
+			log.Err(err), proc.instanceID)
 		return false
 	}
 
 	if !res {
-		logger.With().Info("should not participate: identity is not active",
-			types.LayerID(proc.instanceID))
+		logger.With().Info("should not participate: identity is not active", proc.instanceID)
 		return false
 	}
 
 	currentRole := proc.currentRole(ctx)
 	if currentRole == passive {
 		logger.With().Info("should not participate: passive",
-			log.Uint32("current_k", proc.k), types.LayerID(proc.instanceID))
+			log.Uint32("current_k", proc.getK()), proc.instanceID)
 		return false
 	}
 
+	proc.mu.RLock()
+	eligibilityCount := proc.eligibilityCount
+	proc.mu.RUnlock()
+
 	// should participate
 	logger.With().Info("should participate",
-		log.Uint32("current_k", proc.k), types.LayerID(proc.instanceID),
+		log.Uint32("current_k", proc.getK()), proc.instanceID,
 		log.Bool("leader", currentRole == leader),
-		log.Uint32("eligibility_count", uint32(proc.eligibilityCount)),
+		log.Uint32("eligibility_count", uint32(eligibilityCount)),
 	)
 	return true
 }
 
 // Returns the role matching the current round if eligible for this round, false otherwise
 func (proc *consensusProcess) currentRole(ctx context.Context) role {
 	logger := proc.WithContext(ctx)
-	proof, err := proc.oracle.Proof(ctx, types.LayerID(proc.instanceID), proc.k)
+	proof, err := proc.oracle.Proof(ctx, proc.instanceID, proc.getK())
 	if err != nil {
 		logger.With().Error("could not retrieve eligibility proof from oracle", log.Err(err))
 		return passive
 	}
 
-	eligibilityCount, err := proc.oracle.CalcEligibility(ctx, types.LayerID(proc.instanceID),
-		proc.k, expectedCommitteeSize(proc.k, proc.cfg.N, proc.cfg.ExpectedLeaders), proc.nid, proof)
+	k := proc.getK()
+
+	eligibilityCount, err := proc.oracle.CalcEligibility(ctx, proc.instanceID,
+		k, expectedCommitteeSize(k, proc.cfg.N, proc.cfg.ExpectedLeaders), proc.nid, proof)
 	if err != nil {
-		logger.With().Error("failed to check eligibility", log.Err(err), types.LayerID(proc.instanceID))
+		logger.With().Error("failed to check eligibility", log.Err(err), proc.instanceID)
 		return passive
 	}
 
+	proc.mu.Lock()
 	proc.eligibilityCount = eligibilityCount
+	proc.mu.Unlock()
```

Comment:
also not clear how the eligibility count can be read and updated concurrently

Reply:
```
Write at 0x00c00019e8ea by goroutine 174:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).currentRole()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:903 +0x5a8
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).shouldParticipate()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:864 +0x325
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).beginStatusRound()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:559 +0x2f7
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).onRoundBegin()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:694 +0x8d
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:341 +0xf84
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).Start·dwrap·1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:254 +0x58

Previous write at 0x00c00019e8ea by goroutine 193:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).currentRole()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:903 +0x5a8
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).shouldParticipate()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:864 +0x325
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop.func1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:293 +0x85
```